File: backend/core/api_models/org_api_keys.py
"""
Organization API Keys API Models

Pydantic models for organization-level API key management with scopes.
Supports scopes: read:agents, write:agents, execute:agents, read:templates
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class OrgApiKeyScope(str, Enum):
    """Available scopes for organization API keys"""
    READ_AGENTS = "read:agents"
    WRITE_AGENTS = "write:agents"
    EXECUTE_AGENTS = "execute:agents"
    READ_TEMPLATES = "read:templates"


class OrgApiKeyStatus(str, Enum):
    """Status of an organization API key"""
    ACTIVE = "active"
    REVOKED = "revoked"
    EXPIRED = "expired"


class OrgApiKeyCreateRequest(BaseModel):
    """Request model for creating a new organization API key"""
    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Human-readable name for the API key"
    )
    scopes: List[OrgApiKeyScope] = Field(
        ...,
        min_length=1,
        description="List of scopes for the API key"
    )
    description: Optional[str] = Field(
        None,
        description="Optional description for the API key"
    )
    expires_in_days: Optional[int] = Field(
        None,
        gt=0,
        le=365,
        description="Number of days until expiration (max 365)"
    )

    @field_validator("name")
    def validate_name(cls, v):
        if not v or not v.strip():
            raise ValueError("Name cannot be empty")
        return v.strip()

    @field_validator("scopes")
    def validate_scopes(cls, v):
        if not v:
            raise ValueError("At least one scope is required")
        return list(set(v))  # Remove duplicates


class OrgApiKeyResponse(BaseModel):
    """Response model for API key information (without the secret key)"""
    key_id: UUID
    org_id: UUID
    name: str
    public_key_prefix: str = Field(
        ...,
        description="First 8 characters of public key for identification"
    )
    scopes: List[OrgApiKeyScope]
    description: Optional[str]
    status: OrgApiKeyStatus
    expires_at: Optional[datetime]
    last_used_at: Optional[datetime]
    created_by: UUID
    created_at: datetime


class OrgApiKeyCreateResponse(BaseModel):
    """Response model for newly created API key (includes the full key once)"""
    key_id: UUID
    org_id: UUID
    name: str
    public_key: str = Field(
        ...,
        description="Full public key (opk_xxx format)"
    )
    secret_key: str = Field(
        ...,
        description="Full secret key (osk_xxx format) - only shown once!"
    )
    scopes: List[OrgApiKeyScope]
    description: Optional[str]
    status: OrgApiKeyStatus
    expires_at: Optional[datetime]
    created_by: UUID
    created_at: datetime


class OrgApiKeyListResponse(BaseModel):
    """Response model for listing organization API keys"""
    api_keys: List[OrgApiKeyResponse]
    total: int


class OrgApiKeyValidationResult(BaseModel):
    """Result of organization API key validation"""
    is_valid: bool
    org_id: Optional[UUID] = None
    key_id: Optional[UUID] = None
    scopes: Optional[List[OrgApiKeyScope]] = None
    error_message: Optional[str] = None


class OrgApiKeyUpdateRequest(BaseModel):
    """Request model for updating an organization API key"""
    name: Optional[str] = Field(
        None,
        min_length=1,
        max_length=255,
        description="New name for the API key"
    )
    description: Optional[str] = Field(
        None,
        description="New description for the API key"
    )

    @field_validator("name")
    def validate_name(cls, v):
        if v is not None and not v.strip():
            raise ValueError("Name cannot be empty")
        return v.strip() if v else None

File: apps/frontend/src/lib/api/org-api-keys.ts
/**
 * Organization API Keys API Client
 *
 * API functions for managing organization-level API keys with scopes.
 */

import { backendApi } from "./api";

// Scopes available for organization API keys
export type OrgApiKeyScope =
  | "read:agents"
  | "write:agents"
  | "execute:agents"
  | "read:templates";

export const ORG_API_KEY_SCOPES: { value: OrgApiKeyScope; label: string; description: string }[] = [
  {
    value: "read:agents",
    label: "Read Agents",
    description: "View agents and their configurations",
  },
  {
    value: "write:agents",
    label: "Write Agents",
    description: "Create, update, and delete agents",
  },
  {
    value: "execute:agents",
    label: "Execute Agents",
    description: "Start and stop agent runs",
  },
  {
    value: "read:templates",
    label: "Read Templates",
    description: "View agent templates",
  },
];

export type OrgApiKeyStatus = "active" | "revoked" | "expired";

// Request types
export interface OrgApiKeyCreateRequest {
  name: string;
  scopes: OrgApiKeyScope[];
  description?: string;
  expires_in_days?: number;
}

export interface OrgApiKeyUpdateRequest {
  name?: string;
  description?: string;
}

// Response types
export interface OrgApiKeyResponse {
  key_id: string;
  org_id: string;
  name: string;
  public_key_prefix: string;
  scopes: OrgApiKeyScope[];
  description: string | null;
  status: OrgApiKeyStatus;
  expires_at: string | null;
  last_used_at: string | null;
  created_by: string;
  created_at: string;
}

export interface OrgApiKeyCreateResponse {
  key_id: string;
  org_id: string;
  name: string;
  public_key: string;
  secret_key: string;
  scopes: OrgApiKeyScope[];
  description: string | null;
  status: OrgApiKeyStatus;
  expires_at: string | null;
  created_by: string;
  created_at: string;
}

export interface OrgApiKeyListResponse {
  api_keys: OrgApiKeyResponse[];
  total: number;
}

// API functions

/**
 * Create a new organization API key.
 * The secret key is only returned once upon creation.
 */
export async function createOrgApiKey(
  orgId: string,
  request: OrgApiKeyCreateRequest
): Promise<OrgApiKeyCreateResponse> {
  const response = await backendApi.post<OrgApiKeyCreateResponse>(
    `/organizations/${orgId}/api-keys`,
    request
  );
  return response.data;
}

/**
 * List all API keys for an organization.
 */
export async function listOrgApiKeys(
  orgId: string
): Promise<OrgApiKeyListResponse> {
  const response = await backendApi.get<OrgApiKeyListResponse>(
    `/organizations/${orgId}/api-keys`
  );
  return response.data;
}

/**
 * Get details of a specific API key.
 */
export async function getOrgApiKey(
  orgId: string,
  keyId: string
): Promise<OrgApiKeyResponse> {
  const response = await backendApi.get<OrgApiKeyResponse>(
    `/organizations/${orgId}/api-keys/${keyId}`
  );
  return response.data;
}

/**
 * Update an API key's name or description.
 */
export async function updateOrgApiKey(
  orgId: string,
  keyId: string,
  request: OrgApiKeyUpdateRequest
): Promise<OrgApiKeyResponse> {
  const response = await backendApi.patch<OrgApiKeyResponse>(
    `/organizations/${orgId}/api-keys/${keyId}`,
    request
  );
  return response.data;
}

/**
 * Revoke an API key. The key is deactivated but preserved for audit.
 */
export async function revokeOrgApiKey(
  orgId: string,
  keyId: string
): Promise<{ message: string }> {
  const response = await backendApi.post<{ message: string }>(
    `/organizations/${orgId}/api-keys/${keyId}/revoke`
  );
  return response.data;
}

/**
 * Permanently delete an API key.
 */
export async function deleteOrgApiKey(
  orgId: string,
  keyId: string
): Promise<{ message: string }> {
  const response = await backendApi.delete<{ message: string }>(
    `/organizations/${orgId}/api-keys/${keyId}`
  );
  return response.data;
}

/**
 * Format the full API key for use in requests.
 * Format: {public_key}:{secret_key}
 */
export function formatApiKeyForHeader(
  publicKey: string,
  secretKey: string
): string {
  return `${publicKey}:${secretKey}`;
}

/**
 * Get example curl command for using an API key.
 */
export function getExampleCurlCommand(
  publicKey: string,
  secretKey: string
): string {
  const apiKey = formatApiKeyForHeader(publicKey, secretKey);
  return `curl -H "X-API-Key: ${apiKey}" https://api.example.com/v1/agents`;
}
